src/NumericalAlgorithms/FiniteDifference/Mp5.hpp

```diff
@@ -0,0 +1,174 @@
+             min({abs(w), abs(x), abs(y), abs(z)});
+    };
+
+    // first, compute unlimited fifth-order finite difference reconstruction
+    // values
+    auto result = UnlimitedReconstructor<4>::pointwise(q, stride);
+
+    const double dp = q[2 * stride] + q[0] - 2.0 * q[stride];
+    const double dj = q[stride] + q[-stride] - 2.0 * q[0];
+    const double dm = q[0] + q[-2 * stride] - 2.0 * q[-stride];
+    const double dm4_plus = minmod4(4.0 * dj - dp, 4 * dp - dj, dj, dp);
+    const double dm4_minus = minmod4(4.0 * dj - dm, 4 * dm - dj, dj, dm);
+
+    // compute q_{j+1/2}
+    const double q_mp_plus =
+        q[0] + minmod2(q[stride] - q[0], alpha * (q[0] - q[-stride]));
+
+    if ((result[1] - q[0]) * (result[1] - q_mp_plus) > epsilon) {
+      double q_ul = q[0] + alpha * (q[0] - q[-stride]);
+      double q_md = 0.5 * (q[0] + q[stride] - dm4_plus);
+      double q_lc =
+          q[0] + 0.5 * (q[0] - q[-stride]) + 1.3333333333333333 * dm4_minus;
+      double q_min = max(min({q[0], q[stride], q_md}), min({q[0], q_ul, q_lc}));
+      double q_max = min(max({q[0], q[stride], q_md}), max({q[0], q_ul, q_lc}));
+
+      result[1] = result[1] + minmod2(q_min - result[1], q_max - result[1]);
+    }
+
+    // compute q_{j-1/2}
+    const double q_mp_minus =
+        q[0] + minmod2(q[-stride] - q[0], alpha * (q[0] - q[stride]));
+
+    if ((result[0] - q[0]) * (result[0] - q_mp_minus) > epsilon) {
+      double q_ul = q[0] + alpha * (q[0] - q[stride]);
+      double q_md = 0.5 * (q[0] + q[-stride] - dm4_minus);
+      double q_lc =
+          q[0] + 0.5 * (q[0] - q[stride]) + 1.3333333333333333 * dm4_plus;
+      double q_min =
+          max(min({q[0], q[-stride], q_md}), min({q[0], q_ul, q_lc}));
+      double q_max =
+          min(max({q[0], q[-stride], q_md}), max({q[0], q_ul, q_lc}));
+
+      result[0] = result[0] + minmod2(q_min - result[0], q_max - result[0]);
+    }
+
+    return result;
+  }
+
+  SPECTRE_ALWAYS_INLINE static constexpr size_t stencil_width() { return 5; }
+};
+}  // namespace detail
+
+/*!
+ * \ingroup FiniteDifferenceGroup
+ * \brief Performs the fifth order monotonicity-preserving (MP5) reconstruction
+ * \cite Suresh1997.
+ *
+ * First, calculate the original interface value \f$q_{j+1/2}^\text{OR}\f$ with
+ * the (unlimited) fifth order finite difference reconstruction
+ *
+ * \f{align}
+ *  q_{j+1/2}^\text{OR} = \frac{1}{128}( 3 q_{j-2} - 20 q_{j-1} + 90 q_{j}
+ *            + 60 q_{j+1} - 5 q_{j+2}) .
+ * \f}
+ *
+ * and compute
+ *
+ * \f{align}
+ *  q^\text{MP} = q_j + \text{minmod}(q_{j+1} - q_j, \alpha(q_j - q_{j-1}))
+ * \f}
+ *
+ * for a given value of \f$\alpha\f$ which is usually set to \f$\alpha=4\f$.
+ *
+ * If \f$ (q_{j+1/2}^\text{OR} - q_j)(q_{j+1/2}^\text{OR} - q^\text{MP}) \leq
+ * \epsilon \f$ where \f$\epsilon\f$ is a small tolerance value, use $q_{1/2}
+ * = q_{j+1/2}^\text{OR}$.
+ *
+ * Otherwise, calculate
+ *
+ * \f{align}
+ *  d_j     & = q_{j+1} + q_{j-1} - 2q_j        \\
```

Comment:
These would be a lot easier to understand if they were written always in descending index order instead of `j+1, j-1, j`